modules/common/src/main/scala/com/snowplowanalytics/snowplow/postgres/streaming/KinesisSink.scala

```diff
@@ -0,0 +1,104 @@
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.postgres.streaming
+
+import java.nio.ByteBuffer
+import java.util.UUID
+
+import scala.concurrent.ExecutionContext
+import scala.jdk.FutureConverters._
+import scala.util.{Success, Failure}
+
+import cats.effect.{Async, Resource, Blocker, Sync}
+import cats.implicits._
+
+import fs2.aws.internal.{KinesisProducerClientImpl, KinesisProducerClient}
+
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain
+import com.amazonaws.services.kinesis.producer.UserRecordResult
+import com.amazonaws.services.kinesis.producer.KinesisProducerConfiguration
+import software.amazon.awssdk.services.kinesis.KinesisAsyncClient
+import software.amazon.awssdk.services.kinesis.model.DescribeStreamRequest
+import software.amazon.awssdk.services.kinesis.model.StreamStatus
+import software.amazon.awssdk.regions.Region
+
+import com.google.common.util.concurrent.{Futures, ListenableFuture, FutureCallback}
+
+import com.snowplowanalytics.snowplow.postgres.config.StreamSinkConfig
+
+object KinesisSink {
+
+  def create[F[_]: Async](config: StreamSinkConfig.Kinesis, blocker: Blocker): Resource[F, StreamSink[F]] =
+    mkProducer(config).map(producer => payload => writeToKinesis(payload, config.streamName, producer, blocker))
+
+  private def mkProducer[F[_]: Sync](config: StreamSinkConfig.Kinesis): Resource[F, KinesisProducerClient[F]] =
+    Resource.eval(
+      Sync[F].delay(
+        new KinesisProducerClientImpl[F](Some(mkProducerConfig(config.region.toString)))
+      )
+    )
+
+  private def mkProducerConfig[F[_]](region: String): KinesisProducerConfiguration = {
+    val credentials = new DefaultAWSCredentialsProviderChain()
+    new KinesisProducerConfiguration()
+      .setCredentialsProvider(credentials)
+      .setRegion(region)
+  }
+
+  private def writeToKinesis[F[_]: Async](data: Array[Byte], streamName: String, producer: KinesisProducerClient[F], blocker: Blocker): F[Unit] = {
+    val byteBuffer = ByteBuffer.wrap(data)
+    val partitionKey = UUID.randomUUID().toString
+    producer.putData(streamName, partitionKey, byteBuffer)
+      .flatMap(i => registerCallback(i, blocker)).void
+  }
+
+  private def registerCallback[F[_]: Async](f: ListenableFuture[UserRecordResult], blocker: Blocker): F[UserRecordResult] =
+    Async[F].async[UserRecordResult] { cb =>
+      Futures.addCallback(
+        f,
+        new FutureCallback[UserRecordResult] {
+          override def onFailure(t: Throwable): Unit = cb(Left(t))
```

Comment:
I realized when this case happens, all application crashes since exception isn't handled in anywhere. This means if sending only one event fails, application crashes. Can I ask what is the best practice in here ? Is it okay to let it crash or should we handle it somehow ?

Reply:
What we do in the collector is to keep re-trying these events. A failure is most likely to indicate Kinesis unavailability, and so we keep re-trying until it is back up.

If it's something about the request (eg, too big), then we wouldn't get a `Failure`. We'd get a `Success(response)` and the `response` will detail any problems, including which events from the request were written and which ones were not.

Reply:
That makes sense. I will take a look at collector and try to implement something similar in here also.